build_tools/embed_data/generate_c_embed_data_main.cc

```diff
@@ -0,0 +1,183 @@
+  f << "};\n";
+  GenerateExternCClose(f);
+  f << "#endif  // IREE_FILE_TOC\n";
+}
+
+bool GenerateHeader(const std::string& header_file,
+                    const std::vector<std::string>& toc_files) {
+  std::ofstream f(header_file, std::ios::out | std::ios::trunc);
+  f << "#pragma once\n";  // Pragma once isn't great but is the best we can do.
+  f << "#include <stddef.h>\n";
+  GenerateTocStruct(f);
+  GenerateExternCOpen(f);
+  f << "const struct FileToc* " << absl::GetFlag(FLAGS_identifier)
+    << "_create();\n";
+  f << "static inline size_t " << absl::GetFlag(FLAGS_identifier)
+    << "_size() {\n";
+  f << "  return " << toc_files.size() << ";\n";
+  f << "}\n";
+  GenerateExternCClose(f);
+  f.close();
+  return f.good();
+}
+
+bool SlurpFile(const std::string& file_name, std::string* contents) {
+  constexpr std::streamoff kMaxSize = 100000000;
+  std::ifstream f(file_name, std::ios::in | std::ios::binary);
+  // get length of file:
+  f.seekg(0, f.end);
+  std::streamoff length = f.tellg();
+  f.seekg(0, f.beg);
+  if (!f.good()) return false;
+
+  if (length > kMaxSize) {
+    std::cerr << "File " << file_name << " is too large\n";
+    return false;
+  }
+
+  size_t mem_length = static_cast<size_t>(length);
+  contents->resize(mem_length);
+  f.read(&(*contents)[0], mem_length);
+  f.close();
+  return f.good();
+}
+
+bool GenerateImpl(const std::string& impl_file,
+                  const std::vector<std::string>& input_files,
+                  const std::vector<std::string>& toc_files) {
+  std::ofstream f(impl_file, std::ios::out | std::ios::trunc);
+  f << "#include <stddef.h>\n";
+  f << "#include <stdalign.h>\n";
+  GenerateTocStruct(f);
+  for (size_t i = 0, e = input_files.size(); i < e; ++i) {
+    f << "alignas(alignof(void*)) static char const file_" << i << "[] = {\n";
+    std::string contents;
+    if (!SlurpFile(input_files[i], &contents)) {
+      std::cerr << "Error reading file " << input_files[i] << "\n";
+      return false;
+    }
+    absl::string_view remaining_contents = contents;
+    constexpr int kMaxBytesPerLine = 1024;
+    while (!remaining_contents.empty()) {
+      auto line = remaining_contents.substr(0, kMaxBytesPerLine);
+      f << "\"" << absl::CHexEscape(line) << "\"\n";
+      remaining_contents = remaining_contents.substr(line.size());
+    }
+    f << "};\n";
+  }
+  f << "static const struct FileToc toc[] = {\n";
+  assert(input_files.size() == toc_files.size());
+  for (size_t i = 0, e = input_files.size(); i < e; ++i) {
+    f << "  {\n";
+    f << "    \"" << absl::CEscape(toc_files[i]) << "\",\n";
+    f << "    file_" << i << ",\n";
+    f << "    sizeof(file_" << i << ") - 1\n";
+    f << "  },\n";
+  }
+  f << "  {NULL, NULL, 0},\n";
+  f << "};\n";
+  f << "const struct FileToc* " << absl::GetFlag(FLAGS_identifier)
+    << "_create() {\n";
+  f << "  return &toc[0];\n";
+  f << "}\n";
+
+  f.close();
+  return f.good();
+}
+
+int main(int argc, char** argv) {
```

Comment:
We could merge generate_c_embed_data with generate_cc_embed data to avoid duplicating this generation code. I normally wouldn't be too concerned, but this PR is +500 LOC across multiple files, when it could be leaner.

We could (options 1-3, so one option isn't "C" :P)
(1) Generate two separate .h files for the same .c implementation
(2) Generate one .h file that includes both interfaces.
(3) Drop the C++ part entirely and change all uses to C

Following the general trend of rewriting core code in C, I'm personally a fan of (3), but (2) looks easier to start with

Reply:
How different is the API? I'm also a fan of (3)

Reply:
Chat with @ScottTodd offline. We will go with option (4), same gen tool but output different .h and .c. (2) is really tricky because .c could be used by .cc, so we can't use tricks like `#if __cplusplus`. (3) would require a great clean up of all the namespace usage in the existing test/sample code. We can do that in a followup patch.